frame/uniques/src/lib.rs

```diff
@@ -0,0 +1,1289 @@
+			Ok(())
+		}
+
+		/// Destroy a class of fungible assets.
+		///
+		/// The origin must conform to `ForceOrigin` or must be `Signed` and the sender must be the
+		/// owner of the asset `class`.
+		///
+		/// - `class`: The identifier of the asset class to be destroyed.
+		/// - `witness`: Information on the instances minted in the asset class. This must be
+		/// correct.
+		///
+		/// Emits `Destroyed` event when successful.
+		///
+		/// Weight: `O(n + m)` where:
+		/// - `n = witness.instances`
+		/// - `m = witness.instance_metdadatas`
+		/// - `a = witness.attributes`
+		#[pallet::weight(T::WeightInfo::destroy(
+			witness.instances,
+ 			witness.instance_metadatas,
+			witness.attributes,
+ 		))]
+		pub(super) fn destroy(
+			origin: OriginFor<T>,
+			#[pallet::compact] class: T::ClassId,
+			witness: DestroyWitness,
+		) -> DispatchResult {
+			let maybe_check_owner = match T::ForceOrigin::try_origin(origin) {
+				Ok(_) => None,
+				Err(origin) => Some(ensure_signed(origin)?),
+			};
+			Class::<T, I>::try_mutate_exists(class, |maybe_details| {
+				let class_details = maybe_details.take().ok_or(Error::<T, I>::Unknown)?;
+				if let Some(check_owner) = maybe_check_owner {
+					ensure!(class_details.owner == check_owner, Error::<T, I>::NoPermission);
+				}
+				ensure!(class_details.instances == witness.instances, Error::<T, I>::BadWitness);
+				ensure!(class_details.instance_metadatas == witness.instance_metadatas, Error::<T, I>::BadWitness);
+				ensure!(class_details.attributes == witness.attributes, Error::<T, I>::BadWitness);
+
+				for (instance, details) in Asset::<T, I>::drain_prefix(&class) {
+					Account::<T, I>::remove((&details.owner, &class, &instance));
+				}
+				InstanceMetadataOf::<T, I>::remove_prefix(&class);
+				ClassMetadataOf::<T, I>::remove(&class);
+				Attribute::<T, I>::remove_prefix((&class,));
+				T::Currency::unreserve(&class_details.owner, class_details.total_deposit);
+
+				Self::deposit_event(Event::Destroyed(class));
+
+				// NOTE: could use postinfo to reflect the actual number of accounts/sufficient/approvals
+				Ok(())
+			})
+		}
+
+		/// Mint an asset instance of a particular class.
+		///
+		/// The origin must be Signed and the sender must be the Issuer of the asset `class`.
+		///
+		/// - `class`: The class of the asset to be minted.
+		/// - `instance`: The instance value of the asset to be minted.
+		/// - `beneficiary`: The initial owner of the minted asset.
+		///
+		/// Emits `Issued` event when successful.
+		///
+		/// Weight: `O(1)`
+		#[pallet::weight(T::WeightInfo::mint())]
+		pub(super) fn mint(
+			origin: OriginFor<T>,
+			#[pallet::compact] class: T::ClassId,
+			#[pallet::compact] instance: T::InstanceId,
+			owner: <T::Lookup as StaticLookup>::Source,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			let owner = T::Lookup::lookup(owner)?;
+
+			ensure!(!Asset::<T, I>::contains_key(class, instance), Error::<T, I>::AlreadyExists);
```

Comment:
This pallet and the [pallet-assets](https://github.com/paritytech/substrate/blob/27cc30e88788f6b423d790e34ca39ad0d7fa4219/frame/assets/src/lib.rs#L512-L536) could be more generic if we decouple the token manipulation (transfer, freeze, etc) from the [Monetary Policy](https://github.com/cardano-foundation/docs-cardano-org/blob/main/native-tokens/learn-about-native-tokens.md#important-points-about-minting-policies-and-assets) (mint and burn), the way I see here this pallet only supports the [Single-issuer policy](https://github.com/cardano-foundation/docs-cardano-org/blob/main/native-tokens/learn-about-native-tokens.md#minting-policy-examples), but there's infinite other possible monetary policies (One-time mint, time-locked, bonding curves, atomic swaps, etc), and is important to have those policies enforced on-chain.

I would like to know your opinion about decouple those things, it can be done doing something like this:
``` rust
// Determines whether an account is allowed to mint or burn a particular token.
pub trait MintPolicy<
		ClassId,
		InstanceId,
		Beneficiary,
		MintInput,
		ErrorType
> {
	// - `InstanceId`: The instance value of the asset to be minted
	// - `beneficiary`: The initial owner of the minted asset.
	// - `weight`: weight consumed by mint method
	pub type MintOutput = (InstanceId, Beneficiary, Weight);
	
	fn mint(asset: ClassId, input: MintInput) -> Result<MintOutput, ErrorType>;
}
```
Then
```rust
pub(super) fn mint(
	origin: OriginFor<T>,
	#[pallet::compact] class: T::ClassId,
	input: T::MintInput,
) -> DispatchResult {
	let (instance, owner, consumed) = T::MintPolicy::mint(origin, class, input)?;
	// -- snip
```

Reply:
As the description says, this is meant to be a "basic NFT pallet" which "doesn't try to do anything clever". Your suggestion is therefore beyond the goals of this pallet. As I said in a previous reply to a comment asking for additional features:

> The goals are to provide a minimal possible economically secure NFT implementation. The features provided are really the lowest-common-denominator among NFTs - just enough to implement a basic standard but no more.

> This pallet's functionality could probably be extended with a extension pallet which would add the feaures you want though. If you don't care about the maintenance burden, you could also fork it.